rust/lance/src/index/vector/ivf/builder.rs

```diff
@@ -0,0 +1,215 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+use std::{collections::BTreeMap, ops::Range};
+
+use arrow_array::{cast::AsArray, types::Float32Type, Array, FixedSizeListArray, RecordBatch};
+use arrow_array::{StructArray, UInt32Array};
+use arrow_schema::{DataType, Field, Schema};
+use arrow_select::take::take;
+use futures::{
+    stream::{self, repeat_with},
+    StreamExt, TryStreamExt,
+};
+use lance_arrow::{FixedSizeListArrayExt, RecordBatchExt};
+use lance_linalg::{distance::MetricType, MatrixView};
+use snafu::{location, Location};
+use tracing::instrument;
+
+use crate::dataset::ROW_ID;
+use crate::index::vector::ivf::{
+    io::write_index_partitions, shuffler::ShufflerBuilder, Ivf, PQ_CODE_COLUMN,
+};
+use crate::index::vector::pq::ProductQuantizer;
+use crate::io::object_writer::ObjectWriter;
+use crate::{io::RecordBatchStream, Error, Result};
+
+use super::RESIDUAL_COLUMN;
+
+/// Filter a batch by a range of partition IDs, specified by `part_range`.
+///
+/// Expect the input batch has schema as
+///
+/// ```json
+/// {
+///     "_rowid": "uint32",
+///     "<column>": "fixed_size_list<float16/32/64>",
+/// }
+/// ```
+///
+/// And output batch has schema as
+///
+/// ```json
+/// {
+///     "_rowid": "uint32",
+///     "<column>": "fixed_size_list<float16/32/64>",
+///     "__ivf_part_id": "uint32",
+/// }
+/// ```
+fn filter_batch_by_partition(
+    batch: &RecordBatch,
+    column: &str,
+    ivf: &Ivf,
+    metric_type: MetricType,
+    part_range: Range<u32>,
+) -> Result<Vec<(u32, RecordBatch)>> {
+    let arr = batch
+        .column_by_name(column)
+        .expect("The caller already checked column exist")
+        .as_fixed_size_list();
+    let dim = arr.value_length() as usize;
+
+    let matrix = MatrixView::<Float32Type>::try_from(arr)?;
+    let part_ids = ivf.compute_partitions(&matrix, metric_type);
+    // A map from partition ID and row IDs.
+    let mut parted_map: BTreeMap<u32, Vec<u32>> = BTreeMap::new();
+    part_ids
+        .values()
+        .iter()
+        .enumerate()
+        .for_each(|(idx, &part_id)| {
+            if part_range.contains(&part_id) {
+                parted_map.entry(part_id).or_default().push(idx as u32);
+            }
+        });
+
+    let struct_arr = StructArray::from(batch.clone());
+
+    let residual_field = Field::new(
+        RESIDUAL_COLUMN,
+        DataType::FixedSizeList(
+            Arc::new(Field::new("item", DataType::Float32, true)),
+            dim as i32,
+        ),
+        true,
+    );
+
+    let mut parted_batches = vec![];
+    for (&part_id, row_ids) in parted_map.iter() {
+        let indices = UInt32Array::from(row_ids.clone());
+        // Use `take` to select rows.
+        let str_arr = take(&struct_arr, &indices, None)?;
+        let parted_batch: RecordBatch = str_arr.as_struct().into();
```

Comment:
Don't we have that new `take()` method now?

Reply:
You are right. Fixed.